porch/engine/pkg/engine/patchgen.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package engine
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+
+	api "github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	"github.com/GoogleContainerTools/kpt/porch/repository/pkg/repository"
+	"github.com/bluekeyes/go-gitdiff/gitdiff"
+	"k8s.io/klog/v2"
+
+	"github.com/hexops/gotextdiff"
+	"github.com/hexops/gotextdiff/myers"
+	"github.com/hexops/gotextdiff/span"
+)
+
+func GeneratePatch(fileName string, oldV, newV string) (api.PatchSpec, error) {
+	edits := myers.ComputeEdits(span.URIFromPath(fileName), oldV, newV)
+	diff := fmt.Sprint(gotextdiff.ToUnified(fileName+"@old", fileName+"@new", oldV, edits))
+
+	patchSpec := api.PatchSpec{
+		File:      fileName,
+		PatchType: api.PatchTypePatchFile,
+		Contents:  diff,
+	}
+
+	return patchSpec, nil
+}
+
+type applyPatchMutation struct {
+	patchTask *api.PackagePatchTaskSpec
+	task      *api.Task
+}
+
+var _ mutation = &applyPatchMutation{}
+
+func (m *applyPatchMutation) Apply(ctx context.Context, resources repository.PackageResources) (repository.PackageResources, *api.Task, error) {
+	result := repository.PackageResources{
+		Contents: map[string]string{},
+	}
+
+	for k, v := range resources.Contents {
+		result.Contents[k] = v
+	}
+
+	for _, patchSpec := range m.patchTask.Patches {
+		switch patchSpec.PatchType {
+		case api.PatchTypeCreateFile:
+			if _, found := result.Contents[patchSpec.File]; found {
+				// TODO: We should be able to tolerate this.  Either do a merge or create as a different filename "-2"
+				return result, nil, fmt.Errorf("patch wants to create file %q but already exists", patchSpec.File)
+			}
+			result.Contents[patchSpec.File] = patchSpec.Contents
+		case api.PatchTypeDeleteFile:
+			if _, found := result.Contents[patchSpec.File]; !found {
+				// TODO: I don't think this should be an error, but maybe file manipulation should be replaced with objects
```

Comment:
I think it would make sense to do the patch at the level of resources rather than files

Reply:
I agree ... file patches are more universal (they support comment changes etc), but if we can do an object patch that's much better.  I wanted to start simple though.  I'll add a TODO!